### PR TITLE
🎉 Source Github: bugfix schemas for streams `deployments`, `workflow_runs`, `teams`

### PR DESCRIPTION
**Pull request by grubberr**
Signed-off-by: Sergey Chvalyuk <grubberr@gmail.com>

## What
Try to fix https://github.com/airbytehq/oncall/issues/321#issuecomment-1194525296

1. Improve schemas to conform this validator
https://github.com/airbytehq/airbyte/blob/555efd53604b89e24d746080a546033e00c15329/airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java#L210

2. Improve schemas to conform this transformer
https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py#L34

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by midavadim — APPROVED** [review]

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json
Comment (pedroslopez) [inline airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: From https://github.com/airbytehq/airbyte/issues/15021, `oneOf, anyOf, and not should not be allowed in SAT`

@sherifnada what's the actual effect of doing this?

This should probably just be a `string` type

**Review by pedroslopez — COMMENTED** [review]
Thanks @grubberr - can you confirm you were able to run the traversal against these schemas and it passed?

Also, we should probably hold off on releasing this until the platform side issues from https://github.com/airbytehq/oncall/issues/351 are addressed (see [this comment](https://github.com/airbytehq/oncall/issues/351#issuecomment-1194751389) for more info) - I believe releasing these fixes will cause any schema refresh for github to fail when doing the traversal on the old schema.

File: airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json
Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: Yes, I suspected this question
The problem is if remove this to multi-type, normalization will start to fail (at least destination-snowflake)

**I need your approval to remove this multi-type !!!**

The second point why I have re-implemented using `oneOf`  instead of just simple `["string", "object"]` because it's better handled by python Transformer

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: Other words you can see here just refactoring of previous schema 
but  making it more compatible with python Transformer

Comment (sherifnada) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: @grubberr whats the failure in normalization?

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: something like this
https://github.com/airbytehq/airbyte/issues/15021#issuecomment-1196320638


In short, all multi-type(s) combinations usually stored as pure-JSON on destination 
on some engine it's VARCHAR  but it's not always 

for example:

on snowflake - VARIANT type
on redshift - SUPER type

Comment (sherifnada) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: I suspect the way it was declared made normalization always declare it as an object column. I suspect this means every existing connection is either failing or writing a string? I’m not sure - I haven’t verified this behavior 

The GitHub docs https://docs.github.com/en/rest/deployments/deployments#create-a-deployment seem to indicate it should be a json but still say the api docs say data type could either be Json or string. 

 if this data type is either string or json then it seems from a consumption pov it’s better for this type to always be a json? Otherwise consumers would need to add a json parse which may fail due to not finding a string. 
 
 it also seems like a string is actually technically valid json? But relying on this seems like a hack. So it’s probably best if we make this an object even if the received value is a string by putting that string value inside a “value” key in an object or something? 

@pedroslopez any ideas on how to handle this?

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: > I suspect the way it was declared made normalization always declare it as an object column

I have tested this case: `oneOf(string, object)` and `["string", "object"]` - works the same on normalization
it's stored as JSON object.


About using `{"field": {"type": ["null", "object"]}}` - to cover everything 
it looks like for noramlization it's OK, but for validation it's NOT OK
The problem with this approach that python or java validator don't like it

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: ah sorry I understand now 
you mean conver "string" to `{"value": "string"}`

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: yes I understand your point 
for all such cases where normalization already created JSON-type (redshift-SUPER, snowflake-VARIANT)
cast all to `["null", "object"]` to make things simple 


to pass validation make such tranformations:

```
object -> object
array -> array
"string" -> {"value": "string"}
integer -> {"value": integer}
number -> {"value": "number"}
boolean -> {"value": boolean}
null -> null
```

it looks strange but it can work :)

**Comment by grubberr** [comment]
@pedroslopez 
> can you confirm you were able to run the traversal against these schemas and it passed?

yes I can confirm "Refresh source shema" still failed
it's seems because it failed to validate old schema (0.2.42)

File: airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json
Comment (pedroslopez) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: Declaring our own object format for this (e.g. `{value: "xyz"}`) doesn't feel right to me - I don't think this is something users would expect and they would have to end up dealing with this on their end anyway. My stance on this would be to respect what the source api is emitting as much as possible.

GitHub docs state the schema to be `string ` or `object` in the "response schema" tab of the link sherif shared before:

```
"payload": {
      "oneOf": [
        {
          "type": "object",
          "additionalProperties": true
        },
        {
          "type": "string"
        }
      ]
    },
```

Docs say it should be a JSON text, but If we're not sure whether this string is json, then I think the right thing to do would be to just use the wider type (string in this case).

However, are we actually seeing an issue with the way this is declared now? This seems like a different issue than what's being addressed in the PR.

**Comment by pedroslopez** [comment]
> yes I can confirm "Refresh source shema" still failed it's seems because it failed to validate old schema (0.2.42)

https://github.com/airbytehq/airbyte/pull/15073 was merged and is now being released to cloud which should prevent these errors from happening when comparing with the old schema

File: airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json
Comment (sherifnada) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: the issue is any existing connections seem to have created columns whose type is json/struct (snowflake `variant` or redshift `super`), so trying to ALTER that column to a string is not working

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: couple of notes from my side:

There are 2 variants for this `payload` field

1. with `oneOf`

```json
"payload": {
  "oneOf": [
    {
      "type": "object",
      "properties": {},
      "additionalProperties": true
    },
    {"type": "string"},
    {"type": "null"}
    ]
}
```

2. without `oneOf`

```json
"payload": {
  "type": ["null", "string", "object"]
  "properties": {},
  "additionalProperties": true
}
```

- both variants **work** with normalization for now and produce "JSON"-type on destination
- both variants **work** with validation (but as for me (1) is better that's why I switched to (1))
- both variant **not satisfy new requirements** https://github.com/airbytehq/airbyte/issues/15021

Comment (pedroslopez) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: Ah! Got it - sending a `string` would actually cause an error since they're currently already being handled as `object`s. Then in this case to avoid the breaking schema change, since `payload` is an arbitrary object I think it's ok to make our own object with the string value.

If we confirm that normalization/destinations are just handling this as an object, then we could update the schema here to be `["null", "object"]` (remove string). @grubberr you mentioned earlier that for validation this wouldn't be ok - what issues do we have with validation if we do this?

In practice this should only be:
```
object -> object
"string" -> {"value": "string"}
null -> null
```

(if github docs are true to their word)

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: @pedroslopez if we do such transformation
`"string" -> {"value": "string"}`
no issues with validation

Comment (pedroslopez) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: I asked earlier whether we were seeing an issue with this currently since, while we should fix these multi-type schemas, it seems like a separate issue from what's being addressed in this PR for the on call issue. Since it's on call, I think we should move forward with the current changes and address the schema issue / coercing strings into objects separately

Comment (pedroslopez) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: argh github keeps hiding previous comments before I post 😛 
@grubberr if you have those changes made then feel free to include here or separately, up to you

Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/schemas/deployments.json:118]: @pedroslopez no I did not such changes it's was only my discussion about possible solutions how to move forward to be compliant with this PR https://github.com/airbytehq/airbyte/issues/15021

**Review by pedroslopez — APPROVED** [review]

**Comment by grubberr** [comment]
I see approve from @pedroslopez but @sherifnada did not approved
What we decided ? Can we merge as is for now ?
Thank you

**Review by sherifnada — APPROVED** [review]

**Pull request merged**
